src/main/java/com/minecolonies/coremod/client/gui/WindowPostBox.java

```diff
@@ -0,0 +1,190 @@
+     */
+    private final ScrollingList stackList;
+
+    /**
+     * The filter for the resource list.
+     */
+    private String filter = "";
+
+    /**
+     * The building view of this window.
+     */
+    private AbstractBuildingView buildingView;
+
+    /**
+     * Create the postBox GUI.
+     */
+    public WindowPostBox(final AbstractBuildingView buildingView)
+    {
+        super(buildingView.getID(), Constants.MOD_ID + WINDOW_POSTBOX, buildingView.getColony());
+        this.buildingView = buildingView;
+        this.stackList = findPaneOfTypeByID(LIST_RESOURCES, ScrollingList.class);
+        registerButton(BUTTON_INVENTORY, this::inventoryClicked);
+        registerButton(BUTTON_REQUEST, this::requestClicked);
+    }
+
+    /**
+     * Action executed when request is clicked.
+     * @param button the clicked button.
+     */
+    private void requestClicked(final Button button)
+    {
+        final int row = stackList.getListElementIndexByPane(button);
+        final ItemStack stack = allItems.get(row);
+        int qty = stack.getMaxStackSize();
+        for (final Pane child : button.getParent().getChildren())
+        {
+            if (child.getID().equals(INPUT_QTY))
+            {
+                try
+                {
+                    qty = Integer.parseInt(((TextField) child).getText());
+                    if (qty > stack.getMaxStackSize())
+                    {
+                        qty = stack.getMaxStackSize();
+                    }
+                }
+                catch (final NumberFormatException ex)
+                {
+                    //Be quiet about it.
+                }
+            }
+        }
+        MineColonies.getNetwork().sendToServer(new PostBoxRequestMessage(buildingView, stack, qty));
+    }
+
+    @Override
+    public void onOpened()
+    {
+        super.onOpened();
+        updateResources();
+    }
+
+    /**
+     * Update the item list.
+     */
+    private void updateResources()
+    {
+        final Predicate<ItemStack> filterPredicate = stack -> filter.isEmpty()
+                                                                || stack.getTranslationKey().toLowerCase(Locale.US).contains(filter.toLowerCase(Locale.US))
+                                                                || stack.getDisplayName().toLowerCase(Locale.US).contains(filter.toLowerCase(Locale.US));
+        allItems.clear();
+        allItems.addAll(getBlockList(filterPredicate));
+        updateResourceList();
+    }
+
+    /**
+     * Get the list of blocks which should be added.
+     *
+     * @param filterPredicate the predicate to filter all blocks for.
+     * @return an immutable list of blocks.
+     */
+    private Collection<? extends ItemStack> getBlockList(final Predicate<ItemStack> filterPredicate)
+    {
+        return ImmutableList.copyOf(StreamSupport.stream(Spliterators.spliteratorUnknownSize(Item.REGISTRY.iterator(), Spliterator.ORDERED), false).flatMap(item -> {
```

Comment:
This can be done much more performant.
Instead of iterating over all items in one go, one could possible just collect all items once and then filter them when needed.
This would improve performance in larger modpacks.

Reply:
the problem is that either we do this on open, or we do this on scroll for each scroll action (complex), but at least the moment the player wants to filter in the list it will have to get the whole list

Reply:
Okey.... What ever